cpp/src/table/row_operators.cu

```diff
@@ -402,126 +484,126 @@ namespace {
   std::vector<std::unique_ptr<column>> out_cols_rhs;
 
+  auto const make_output = [&](auto const& new_child_lhs, auto const& new_child_rhs) {
+    return std::tuple{replace_child(lhs, new_child_lhs, out_cols_lhs, stream, mr),
+                      replace_child(rhs, new_child_rhs, out_cols_rhs, stream, mr),
+                      std::move(out_cols_lhs),
+                      std::move(out_cols_rhs)};
+  };
+
   if (lhs.type().id() == type_id::LIST) {
     auto const child_lhs = cudf::lists_column_view{lhs}.get_sliced_child(stream);
+    auto const child_rhs = cudf::lists_column_view{rhs}.get_sliced_child(stream);
 
     // Found a lists-of-structs column.
     if (child_lhs.type().id() == type_id::STRUCT) {
-      if (rhs_opt) {  // rhs table is available
-        auto const child_rhs = cudf::lists_column_view{rhs_opt.value()}.get_sliced_child(stream);
-        auto const concatenated_children = cudf::detail::concatenate(
-          std::vector<column_view>{child_lhs, child_rhs}, stream, default_mr);
-
-        auto const ranks        = compute_ranks(concatenated_children->view());
-        auto const ranks_slices = cudf::detail::slice(
-          ranks->view(),
-          {0, child_lhs.size(), child_lhs.size(), child_lhs.size() + child_rhs.size()},
-          stream);
-
-        out_cols_lhs.emplace_back(std::make_unique<column>(ranks_slices.front()));
-        out_cols_rhs.emplace_back(std::make_unique<column>(ranks_slices.back()));
-
-        auto transformed_lhs = replace_child(lhs, out_cols_lhs.back()->view(), out_cols_lhs);
-        auto transformed_rhs =
-          replace_child(rhs_opt.value(), out_cols_rhs.back()->view(), out_cols_rhs);
-
-        return {std::move(transformed_lhs),
-                std::optional<column_view>{std::move(transformed_rhs)},
-                std::move(out_cols_lhs),
-                std::move(out_cols_rhs)};
-      } else {  // rhs table is not available
-        out_cols_lhs.emplace_back(compute_ranks(child_lhs));
-        auto transformed_lhs = replace_child(lhs, out_cols_lhs.back()->view(), out_cols_lhs);
-
-        return {std::move(transformed_lhs),
-                std::nullopt,
-                std::move(out_cols_lhs),
-                std::move(out_cols_rhs)};
-      }
+      auto const concatenated_children =
+        cudf::detail::concatenate(std::vector<column_view>{child_lhs, child_rhs},
+                                  stream,
+                                  rmm::mr::get_current_device_resource());
+
+      auto const ranks =
+        compute_ranks(concatenated_children->view(), column_null_order, stream, mr);
+      auto const ranks_slices = cudf::detail::slice(
+        ranks->view(),
+        {0, child_lhs.size(), child_lhs.size(), child_lhs.size() + child_rhs.size()},
+        stream);
+
+      out_cols_lhs.emplace_back(std::make_unique<column>(ranks_slices.front()));
+      out_cols_rhs.emplace_back(std::make_unique<column>(ranks_slices.back()));
```

Comment:
This makes a copy so the stream should be passed
```suggestion
      out_cols_lhs.emplace_back(std::make_unique<column>(ranks_slices.front(), stream));
      out_cols_rhs.emplace_back(std::make_unique<column>(ranks_slices.back(), stream));
```

Reply:
Oh didn't realize that. Also need to pass `mr`.